worker-macros/src/durable_object.rs

```diff
@@ -109,24 +124,123 @@ pub fn expand_macro(tokens: TokenStream) -> syn::Result<TokenStream> {
+
+                        quote! {
+                            #pound[wasm_bindgen::prelude::wasm_bindgen(js_name = webSocketError)]
+                            pub fn _on_error(&mut self, ws: worker_sys::web_sys::WebSocket, error: JsValue) -> js_sys::Promise {
+                                // SAFETY:
+                                // On the surface, this is unsound because the Durable Object could be dropped
+                                // while JavaScript still has possession of the future. However,
+                                // we know something that Rust doesn't: that the Durable Object will never be destroyed
+                                // while there is still a running promise inside of it, therefore we can let a reference
+                                // to the durable object escape into a static-lifetime future.
+                                let static_self: &'static mut Self = unsafe {&mut *(self as *mut _)};
+
+                                wasm_bindgen_futures::future_to_promise(async move {
+                                    static_self._on_error_raw(ws.into(), error.into()).await.map(|_| wasm_bindgen::JsValue::NULL)
+                                        .map_err(wasm_bindgen::JsValue::from)
+                                })
+                            }
+
+                            #method
+                        }
+                    },
                     _ => panic!()
                 };
                 tokenized.push(tokens);
             }
 
-            let alarm_tokens = has_alarm.then(|| quote! {
+            let alarm_tokens = optional_methods.has_alarm.then(|| quote! {
                 async fn alarm(&mut self) -> ::worker::Result<worker::Response> {
                     self._alarm_raw().await
                 }
             });
+            
+            let on_message_tokens = optional_methods.has_on_message.then(|| quote! {
+                async fn on_message(&mut self, ws: ::worker::WebSocket, message: String) -> ::worker::Result<()> {
+                    self._on_message_raw(ws, message).await
+                }
+            });
+
+            let on_close_tokens = optional_methods.has_on_close.then(|| quote! {
+                async fn on_close(&mut self, ws: ::worker::WebSocket, code: usize, reason: String, was_clean: bool) -> ::worker::Result<()> {
+                    self._on_close_raw(ws, code, reason, was_clean).await
+                }
+            });
+
+            let on_error_tokens = optional_methods.has_on_error.then(|| quote! {
+                async fn on_error(&mut self, ws: ::worker::WebSocket, error: ::worker::Error) -> ::worker::Result<()> {
+                    self._on_error_raw(ws, error).await
+                }
+            });
+
             Ok(quote! {
                 #wasm_bindgen_attr
                 impl #struct_name {
                     #(#tokenized)*
                 }
 
-                #pound[async_trait::async_trait(?Send)]
```

Comment:
`async_trait` isn't required anymore, but if you wanted to still support older versions of Rust I can add it back in.